plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -484,53 +510,62 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
             }
          };
 
-         const auto& id = trx->id();
-         if( fc::time_point(trx->packed_trx()->expiration()) < block_time ) {
-            send_response(std::static_pointer_cast<fc::exception>(
-                  std::make_shared<expired_tx_exception>(
-                        FC_LOG_MESSAGE(error, "expired transaction ${id}, expiration ${e}, block time ${bt}",
-                                       ("id", id)("e", trx->packed_trx()->expiration())("bt", block_time)) )));
-            return;
-         }
+         try {
+            if( !chain.is_building_block()) {
+               _pending_incoming_transactions.add( trx, persist_until_expired, next );
+               return;
+            }
 
-         if( chain.is_known_unexpired_transaction(id) ) {
-            send_response(std::static_pointer_cast<fc::exception>(std::make_shared<tx_duplicate>(FC_LOG_MESSAGE(error, "duplicate transaction ${id}", ("id", id)) )));
-            return;
-         }
+            auto block_time = chain.pending_block_time();
 
-         auto deadline = fc::time_point::now() + fc::milliseconds(_max_transaction_time_ms);
-         bool deadline_is_subjective = false;
-         const auto block_deadline = calculate_block_deadline(block_time);
-         if (_max_transaction_time_ms < 0 || (_pending_block_mode == pending_block_mode::producing && block_deadline < deadline) ) {
-            deadline_is_subjective = true;
-            deadline = block_deadline;
-         }
+            const auto& id = trx->id();
+            if( fc::time_point( trx->packed_trx()->expiration()) < block_time ) {
```

Comment:
we used to do this check first, is there concern that `block_time` may move backwards? otherwise this seems like it may store an expired transaction in the `_pending_incoming_transaction` queue

Reply:
oh, I'm incorrect it was always in this order.  But perhaps we should swap these now such that we check expiration before storing on the pending list.

Same with other early outs like `is_known_unexpired_transaction` .  I am relatively sure that even when faced with forking conditions any transaction that returns `true` will end up on the unapplied transactions list and therefore not need to be duplicated in our pending queue.

Reply:
I left it that way because `pending_block_time` is not available if `!chain.is_building_block()`. Changed it to use either pending block time or head block time.